File: 01-ordenacao/java/OrdenacaoMVC/src/model/algoritmos/SelectionSort.java
package model.algoritmos;

import model.ResultadoOrdenacao;
import java.util.List;

/**
 * Selection Sort
 * Seleciona o menor do trecho restante e coloca na posição correta
 */
public class SelectionSort implements IAlgoritmoOrdenacao {

    @Override
    public String getNome() {
        return "Selection Sort";
    }

    @Override
    public ResultadoOrdenacao ordenar(List<Integer> lista) {
        long comparacoes = 0;
        long trocas = 0;
        int posMenor;
        int tmp;

        long inicio = System.nanoTime();

        for (int i = 0; i < lista.size() - 1; i++) {
            posMenor = i;
            for (int j = i + 1; j < lista.size(); j++) {
                comparacoes++;
                if (lista.get(j) < lista.get(posMenor)) {
                    posMenor = j;
                }
            }
            if (i != posMenor) {
                trocas++;
                tmp = lista.get(i);
                lista.set(i, lista.get(posMenor));
                lista.set(posMenor, tmp);
            } // troca só se a posição mudou (evita contar troca consigo mesmo)
        }

        long tempoMs = (System.nanoTime() - inicio) / 1_000_000;

        return new ResultadoOrdenacao(comparacoes, trocas, 0, tempoMs, lista.size());
    }
}

File: 01-ordenacao/csharp/OrdenacaoMVC/src/Model/Algoritmos/SelectionSort.cs
using System.Diagnostics;

namespace OrdenacaoMVC.Model.Algoritmos;

/// <summary>
/// Implementação do Selection Sort (ordenação por seleção)
/// A cada iteração seleciona o menor elemento do trecho restante e o coloca na posição correta
///</summary>
public class SelectionSort : IAlgoritmoOrdenacao
{
    public string Nome => "Selection Sort";

    public ResultadoOrdenacao Ordenar(List<int> lista)
    {
        long comparacoes = 0;
        long trocas = 0;
        int posMenor;
        int tmp;

        Stopwatch sw = Stopwatch.StartNew();

        for (int i = 0; i < lista.Count - 1; i++)
        {
            posMenor = i;
            for (int j = i + 1; j < lista.Count; j++)
            {
                comparacoes++;
                if (lista[j] < lista[posMenor])
                {
                    posMenor = j;
                }
            }
            if (i != posMenor)
            {
                trocas++;
                tmp = lista[i];
                lista[i] = lista[posMenor];
                lista[posMenor] = tmp;
            }
        }

        sw.Stop();

        return new ResultadoOrdenacao(
            Comparacoes: comparacoes,
            Trocas: trocas,
            Movimentacoes: 0,
            TempoMs: sw.ElapsedMilliseconds,
            TamanhoLista: lista.Count);
    }
}
